### PR TITLE
feat: clarify the source of lifecycle hook errors

### PR DESCRIPTION
**Pull request by kormide**
Addresses https://github.com/aspect-build/rules_js/issues/494.

fyi @Silic0nS0ldier

This doesn't present any new information, just makes it a bit easier to visually parse the build log failure when a lifecycle script fails.

e.g.
```
ERROR: /home/derek/Code/rules_js/BUILD.bazel:10:22: Running lifecycle hooks on npm package @aspect-test/c@2.0.2 failed: (Exit 1): lifecycle-hooks.sh failed: error executing command (from target //:.aspect_rules_js/node_modules/@aspect-test+c@2.0.2/lc) bazel-out/k8-opt-exec-2B5CBBC6/bin/npm/private/lifecycle/lifecycle-hooks.sh @aspect-test/c ../../../external/npm__at_aspect-test_c__2.0.2/package ... (remaining 1 argument skipped)

> @aspect-test/c@2.0.2 postinstall /home/derek/.cache/bazel/_bazel_derek/8fc6941ac57b91558bc3587efd7d8c3c/execroot/aspect_rules_js/bazel-out/k8-fastbuild/bin/node_modules/.aspect_rules_js/@aspect-test+c@2.0.2/node_modules/@aspect-test/c
> bash -c 'echo {\"answer\":\"42*\"}' > data.json


> @ custom_postinstall /home/derek/.cache/bazel/_bazel_derek/8fc6941ac57b91558bc3587efd7d8c3c/execroot/aspect_rules_js/bazel-out/k8-fastbuild/bin/node_modules/.aspect_rules_js/@aspect-test+c@2.0.2/node_modules/@aspect-test/c
> exit 1 && echo moo > cow.txt && echo mooo >> cow.txt

===============================================================
Failure while running lifecycle hook for package '@':

  Script:  'custom_postinstall'
  Command: `exit 1 && echo moo > cow.txt && echo mooo >> cow.txt`

Stack trace:

Exit status 1
    at EventEmitter.<anonymous> (/home/derek/.cache/bazel/_bazel_derek/8fc6941ac57b91558bc3587efd7d8c3c/execroot/aspect_rules_js/bazel-out/k8-fastbuild/bin/npm/private/lifecycle/min/index.min.js:1:73836)
    at EventEmitter.emit (node:events:394:28)
    at ChildProcess.<anonymous> (/home/derek/.cache/bazel/_bazel_derek/8fc6941ac57b91558bc3587efd7d8c3c/execroot/aspect_rules_js/bazel-out/k8-fastbuild/bin/npm/private/lifecycle/min/index.min.js:1:80298)
    at ChildProcess.emit (node:events:394:28)
    at maybeClose (node:internal/child_process:1064:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
===============================================================
```

### CODE REVIEW

File: npm/private/lifecycle/lifecycle-hooks.js [hunk 1]
```diff
@@ -167,7 +167,19 @@ async function copyRecursive(src, dest) {
...
+            console.log("===============================================================")
+            console.log(`Failure while running lifecycle hook for package '${e.pkgid}':\n`);
+            console.log(`  Script:  '${e.stage}'`);
+            console.log(`  Command: \`${e.script}\``);
+            console.log(`\nStack trace:\n`);
+            // First line of error is always the message, which is redundant with the above logging.
+            console.log(e.stack.replace(/^.*?\n/, ""));
+            console.log("===============================================================");
+        } else {
+            console.error(e);
```
Comment (kormide) [inline npm/private/lifecycle/lifecycle-hooks.js:180]: `console.log` produces a more logical output order, as pointed out [here](https://github.com/aspect-build/rules_js/issues/217#issuecomment-1268128063).

File: npm/private/lifecycle/lifecycle-hooks.js [hunk 2]
```diff
@@ -167,7 +167,19 @@ async function copyRecursive(src, dest) {
     try {
         await main(process.argv.slice(2))
     } catch (e) {
-        console.error(e)
+        if (e.code === "ELIFECYCLE" && e.pkgid && e.stage && e.script) {
```
Comment (alexeagle) [inline npm/private/lifecycle/lifecycle-hooks.js:170]:
```suggestion
        if (e.code === "ELIFECYCLE" && !!e.pkgid && !!e.stage && !!e.script) {
```

File: npm/private/lifecycle/lifecycle-hooks.js [hunk 3]
```diff
@@ -167,7 +167,19 @@ async function copyRecursive(src, dest) {
...
+            console.log("===============================================================")
```
Comment (alexeagle) [inline npm/private/lifecycle/lifecycle-hooks.js:171]: how come these go to stdout rather than stderr? is there some reasoning for that?

Comment (alexeagle) [reply npm/private/lifecycle/lifecycle-hooks.js:171]: oh I see, you commented earlier. Consider leaving a comment in the code so someone doesn't change it later.

**Review by alexeagle — APPROVED** [review]

File: npm/private/lifecycle/lifecycle-hooks.js (hunk 3 above)
Comment (kormide) [reply npm/private/lifecycle/lifecycle-hooks.js:173]: Added a comment.

**Pull request merged**